docs/design/cpsubsystem/04-cp-object-lifecycle.md

```diff
@@ -0,0 +1,319 @@
+### Background
+#### Motivation
+
+The current lifecycle of a CP object does not allow the same object name to be reused after it was destroyed.
+
+```java
+IAtomicLong atomicLong = cpSubsystem.getAtomicLong("iLong");
+atomicLong.set(2);
+atomicLong.destroy();
+
+IAtomicLong atomicLongNew = cpSubsystem.getAtomicLong("iLong");
+atomicLongNew.get(); // throws DistributedObjectDestroyedException
+```
+
+This leads to:
+- memory leaks (non-significant) since the CP subsystem should track the names of all destroyed objects
+- customers dissatisfaction since they should generate another name of the object
+- customers do not clean up unnecessary objects to avoid creating another object name, which results in additional overhead and memory consumption
+
+But trying to introduce the ability to reuse the name of a destroyed object, based on the existing "get-or-create" semantics for CP objects, will allow the following behaviour:
+```java
+IAtomicLong atomicLong = cpSubsystem.getAtomicLong("iLong");
+atomicLong.set(2);
+atomicLong.destroy();
+atomicLong.get(); // succeeds and implicitly recreates internal atomicLong state
+```
+
+Furthermore
+
+<table>
+<tr>
+<td> Client 1 </td> <td> Client 2 </td>
+</tr>
+<tr>
+<td> 
+
+```java
+FencedLock lock = cpSubsystem.getLock("iLock");
+lock.lock();
+lock.destroy();
+```
+
+</td>
+<td>
+
+```java
+FencedLock lock = cpSubsystem.getLock("iLock");
+
+
+lock.lock(); // recreate and take the lock
+```
+
+</td>
+</tr>
+</table>
+
+Such behaviour breaks the linearizable CP object history and not all clients of object may be aware of the object destruction.
+
+The desirable behaviour, corresponding to a linearizable history, should be as follows:
+
+```java
+IAtomicLong atomicLong = cpSubsystem.getAtomicLong("iLong");
+atomicLong.set(2);
+atomicLong.destroy();
+atomicLong.get(); // should failed with exception
+
+atomicLong = cpSubsystem.getAtomicLong("iLong"); // should demand obtaining a fresh proxy
+atomicLong.get(); // should succeed
+```
+
+#### Current CP object lifecycle implementation
+
+All client interactions with CP objects take place through a proxy: Client proxy or Local proxy in an embedded mode. 
+
+Calling the `CPSubsystem#getLock/getAtomicLong/…` method returns a proxy object and does not trigger a real object creation/initialisation in a CP Subsystem.
+
+An actual object creation/initialisation happens in a lazy way by any business method call on a proxy (for example, `atomicLong.get()`):
+
+```java
+V atomicValue = atomicValues.get(key);
+if (atomicValue == null) {
+    atomicValue = newAtomicValue(groupId, name, null);
+    atomicValues.put(key, atomicValue);
+}
+```
+
+Thus, if an object was destroyed, a new object will be created.
+
+--Sequence diagram--
```

Comment:
Did you mean to add another diagram here?

Reply:
yes (still in progress ...)